2_process/src/add_dynamic_attr.R

```diff
@@ -0,0 +1,796 @@
+  #' This is used to compute the lagged attributes
+  #' @param attr_name character string of a name in the attribute column of lag_table
+  #' @param start_date first date to compute attributes
+  #' @param segs vector of the PRMS reach IDs
+  #' 
+  #' @return tbl with columns for Date and seg
+  
+  #Get the lags for this attribute
+  tmp_lags <- lag_table %>% 
+    filter(attribute == attr_name) %>%
+    pull(lags) %>%
+    unlist()
+  tmp_lag_unit <- lag_table %>% 
+    filter(attribute == attr_name) %>%
+    pull(lag_unit) %>%
+    unlist()
+  
+  #Make the length of the lag_unit match the length of lags
+  if ((length(tmp_lag_unit) == 1) & (length(tmp_lags) > 1)){
+    tmp_lag_unit <- rep(tmp_lag_unit, length(tmp_lags))
+  }
+  
+  #Find the earliest date: (start_date - largest lag)
+  for (lag in 1:length(tmp_lags)){
+    if(lag == 1){
+      min_lag_date <- as.Date(start_date) %m-% 
+        period(paste(tmp_lags[lag], tmp_lag_unit[lag]))
+    }else{
+      min_lag_date <- min(min_lag_date, 
+                          as.Date(start_date) %m-% 
+                            period(paste(tmp_lags[lag], tmp_lag_unit[lag])))
+    }
+  }
+  
+  #Create the tbl
+  dates <- seq(min_lag_date, as.Date(start_date) %m-% period(1, 'day'), 1)
+  df <- tibble(Date = dates, seg = segs[1])
+  for(i in 2:length(segs)){
+    df_tmp <- tibble(Date = dates, seg = segs[i])
+    df <- rbind(df, df_tmp)
+  }
+  
+  return(df)
+}
+
+
+compute_lagged_attrs <- function(dyn_df, lag_table, lag_attr_name, 
+                                 attr_name, start_date){
+  #' @description computes lagged predictor variables from the provided features
+  #' 
+  #' @param dyn_df tbl containing "Date" and the PRMS "seg". Should have all
+  #' dates that you want to use for lagged computations.
+  #' @param lag_table table with 3 columns: attributes, lags, and lag_unit.
+  #' This is used to compute the lagged attributes
+  #' @param lag_attr_name name in the attribute column of lag_table to use
+  #' @param attr_name name to use for this attribute. Format of the new name will
+  #' be paste0(attr_prefix, '_', attr_name, '_', lags, lag_unit)
+  #' @param start_date first date to compute attributes
+  #' 
+  #' @return tbl with the added lagged features. Date begins at the start_date
+  #' and ends at the latest provided date in dyn_df.
+  
+  #Get the lag information and lag functions to be used
+  tmp_lags <- lag_table %>% 
+    filter(attribute == lag_attr_name) %>%
+    pull(lags) %>%
+    unlist()
+  tmp_lag_unit <- lag_table %>% 
+    filter(attribute == lag_attr_name) %>%
+    pull(lag_unit) %>%
+    unlist()
+  #Make the length of the lag_unit match the length of lags
+  if ((length(tmp_lag_unit) == 1) & (length(tmp_lags) > 1)){
+    tmp_lag_unit <- rep(tmp_lag_unit, length(tmp_lags))
+  }
+  tmp_lag_fxns <- lag_table %>% 
+    filter(attribute == lag_attr_name) %>%
+    pull(lag_fxns) %>%
+    unlist()
```

Comment:
That's a good catch. I initially had the `exact` computation as an `else` in this, and then decided to move it because it was faster to compute outside of this function.